runtime/runtime/src/congestion_control.rs

```diff
@@ -0,0 +1,388 @@
+        // limit" is the safest approach to ensure availability.
+        let forward_limit = outgoing_limit.entry(shard).or_insert(Gas::MAX);
+        let gas_to_forward = receipt_congestion_gas(&receipt, &apply_state.config)?;
+        if *forward_limit > gas_to_forward {
+            outgoing_receipts.push(receipt);
+            // underflow impossible: checked forward_limit > gas_to_forward above
+            *forward_limit -= gas_to_forward;
+            Ok(ReceiptForwarding::Forwarded)
+        } else {
+            Ok(ReceiptForwarding::NotForwarded(receipt))
+        }
+    }
+
+    /// Put a receipt in the outgoing receipt buffer of a shard.
+    fn buffer_receipt(
+        &mut self,
+        receipt: &Receipt,
+        state_update: &mut TrieUpdate,
+        shard: u64,
+        config: &RuntimeConfig,
+    ) -> Result<(), RuntimeError> {
+        let bytes = receipt_size(&receipt)?;
+        let gas = receipt_congestion_gas(&receipt, config)?;
+        self.congestion_info.add_receipt_bytes(bytes as u64)?;
+        self.congestion_info.add_buffered_receipt_gas(gas)?;
+        self.outgoing_buffers.to_shard(shard).push(state_update, &receipt)?;
+        Ok(())
+    }
+}
+
+fn receipt_congestion_gas(
+    receipt: &Receipt,
+    config: &RuntimeConfig,
+) -> Result<Gas, IntegerOverflowError> {
+    match &receipt.receipt {
+        ReceiptEnum::Action(action_receipt) => {
+            // account for gas guaranteed to be used for executing the receipts
+            let prepaid_exec_gas = safe_add_gas(
+                total_prepaid_exec_fees(config, &action_receipt.actions, &receipt.receiver_id)?,
+                config.fees.fee(ActionCosts::new_action_receipt).exec_fee(),
+            )?;
+            // account for gas guaranteed to be used for creating new receipts
+            let prepaid_send_gas = total_prepaid_send_fees(config, &action_receipt.actions)?;
+            let prepaid_gas = safe_add_gas(prepaid_exec_gas, prepaid_send_gas)?;
+
+            // account for gas potentially used for dynamic execution
+            let gas_attached_to_fns = total_prepaid_gas(&action_receipt.actions)?;
+            let gas = safe_add_gas(gas_attached_to_fns, prepaid_gas)?;
+
+            Ok(gas)
+        }
+        ReceiptEnum::Data(_data_receipt) => {
+            // Data receipts themselves don't cost gas to execute, their cost is
+            // burnt at creation. What we should count, is the gas of the
+            // postponed action receipt. But looking that up would require
+            // reading the postponed receipt from the trie.
+            // Thus, the congestion control MVP does not account for data
+            // receipts or postponed receipts.
+            Ok(0)
+        }
+        ReceiptEnum::PromiseYield(_) => {
+            // The congestion control MVP does not account for yielding a promise.
+            // It does not burn any gas, so it's unclear what we could account for.
```

Comment:
Thanks for pointing this out. @nagisa  But I still struggle to understand the correct way of accounting gas when a `PromiseYield` receipt is in the delayed receipts queue or an outgoing buffer.

For context, we are trying to estimate the work to process this receipt when it enters `apply` from the delayed receipts queue. This is the congestion signal we care about. If the work is not ready to execute now, it must be ignored or else we run into the risk of work starvation of shards. Imagine we fill all buffers with yielded promises, now we will stop accepting receipts, including PromiseResume messages. We would be deadlocked until a timeout makes space again.

That's why I thought for `PromiseYield` we should ignore the attached gas. And for `PromiseResume` we *should theoretically* account for all the gas in the resumed promise, but that is not known without looking up the receipt first. So we ignore that, too for lack of better alternative.

As I understand it, creation of the receipt is already covered in contract execution using the attached gas there. So this seems quite clearly not something we need to account for anymore.

And next, I thought we just put `PromiseYield` in a separate column of the local state trie. At this moment, it does not contribute to congestion. In other words, it does not take space in outgoing buffers or delayed receipts queues. Like postponed receipts, we can ignore it.

Now that I think about it, is there even a way how a `ReceiptEnum::PromiseYield` could ever be forwarded to another shard or somehow else show up in the delayed receipt queue? I think only `PromiseResume` could show up, right? If that is the case, I think I should update the comments to explain that the case is irrelevant as it can't happen in practice.

Reply:
> is there even a way how a ReceiptEnum::PromiseYield could ever be forwarded to another shard

No. Anything involving yielded promises, including resumes, is not only confined to the same shard, but also to the same account. At least currently.

> in the delayed receipt queue?

Now this one I don't know, unfortunately. These promises are acted upon separately from delayed receipts, but due to how transaction runtime is structured I think we might never get to these yields if the shard is perpetually congested and is busy with delayed receipts all the time.

---

If this only concerns itself with delayed receipts then this part of code is indeed dead to the best of my knowledge. Might make sense to log an error or something, so this MVPness of the implementation doesn't fly under the radar if our assumptions happen to be wrong somehow.